perf(stdlib): optimize pivot for common scenarios

jsternberg opened the pull request:
This optimizes pivot for the common scenario of a single row and single
column key where the column key is part of the group key.

It will perform a pivot by reading all of the relevant tables and it
will then merge all of the keys together in sorted order. After it has
done that, it will build each column by matching the value with the
appropriate key. This causes the algorithm to run in `O(n)` time and
avoid unnecessary reallocations because it determines the size before it
attempts to build any arrays.

#2133

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

jsternberg commented:
Benchmark for this:

```
jsternberg:[~/g/s/g/i/flux] $ (perf/optimize-pivot *$) benchcmp old.txt new.txt
benchmark                  old ns/op       new ns/op      delta
BenchmarkPivot/1000-12     33557264224     5613306182     -83.27%

benchmark                  old allocs     new allocs     delta
BenchmarkPivot/1000-12     136673348      3813245        -97.21%

benchmark                  old bytes      new bytes      delta
BenchmarkPivot/1000-12     9367265416     3319302120     -64.56%
```

It likely still needs some tests, but this PR doesn't enable the code. I think we're likely going to need #2131 for this to work effectively so we can confidently enable this code in specific scenarios.

jsternberg commented:
I updated this so it works with the new way that tools are versioned. This is ready for review again.

jsternberg commented:
Test cases from the v1 version of pivot were adapted for v2. The v2 one only works if the column key is part of the group key and that the row key has no nulls so those tests were removed.

nathanielc commented on stdlib/universe/pivot.go:
```diff
@@ -382,3 +390,285 @@ func (t *pivotTransformation) Finish(id execute.DatasetID, err error) {
+}
+
+//go:generate -command tmpl ../../gotool.sh github.com/benbjohnson/tmpl
+//go:generate tmpl -data=@../../internal/types.tmpldata -o pivot.gen.go pivot.gen.go.tmpl
```
nit, can you move these to the top of the file?

jsternberg replied:
Sure. I put it at the bottom because that's where the code originally was, but it makes sense at the top too.

jsternberg commented:
I think I probably do need to create tests for the new functionality that isn't being tested. I'll try to write those today. If this begins to get too delayed because I'm having difficulty writing the tests, we can split the work.

jsternberg commented:
Wrote tests for arrowutil and the other stuff that's in here. I also cleaned up the iterators a bit. Previously, I had next be called after reading the value, but that's not really expected in typical Go APIs. I updated it so next is always called before reading the value. That made the merge code a bit different (and the previous api made it easier), but I don't think there's much of a meaningful difference and having the iterators with a weird API specifically for one algorithm seemed a bit strange to me.

jsternberg commented:
I tried to add more patch coverage. I think part of the issue is a lot of the generated file is not actually tested because it generates the code paths for the row key to be a non-time column. I think I am just going to have to go with what I currently have for now. I believe the test coverage to be sufficient enough.
